docs/enhancements/dynamic-fetching-of-openshift-versions.md

```diff
@@ -0,0 +1,289 @@
+
+### User Stories
+
+#### Installing The Latest Version of a z-stream
+
+As a user of assisted-installer UI, I should have a list of the latest versions
+for each minor release.
+
+#### Installing an Older Version of a z-stream
+
+As a user of assisted-installer UI, I should have the option to browse the
+different available versions that are fetched from upgrade channels.
+
+#### Configuring the Dynamically-loaded Versions of a Stream
+
+As a system administrator for the assisted-installer, I should have the option
+to configure both dynamically-loaded and statically-loaded OpenShift version.
+To do that, I can provide the following environment variables for example:
+```
+RELEASE_IMAGES='[
+  {
+    "openshift_version": "4.12-multi",
+    "cpu_architecture": "multi",
+    "cpu_architectures": [
+        "x86_64",
+        "arm64",
+        "ppc64le",
+        "s390x"
+    ],
+    "url": "quay.io/openshift-release-dev/ocp-release:4.12.0-rc.4-multi",
+    "version": "4.12.0-rc.4"
+  }
+]'
+
+DYNAMIC_OPENSHIFT_RELEASES_CONFIG='[
+  {
+    "openshift_version": "4.11",
+    "upgrade_channels": [
+        {
+          "cpu_architecture": "x86_64",
+          "channels": ["stable"]
+        }
+    ]
+  },
+  {
+    "openshift_version": "4.11",
+    "upgrade_channels": [
+      {
+        "cpu_architecture": "x86_64",
+        "channels": ["candidate", "stable"]
+      },
+      {
+        "cpu_architecture": "arm64",
+        "channels": ["candidate", "stable"]
+      }
+    ]
+  }
+]'
+```
+
+### Risks and Mitigations
+
+We assume here that each x.y RHCOS version is backwards and forwards compatible
+with all x.y.z OpenShift versions.
+
+We have recently seen occasions where the disk space is being exhausted due to
+the assisted-service caching a lot of different openshift installer binaries.
+It is important to make sure we don't make this problem more severe, as this
+enhancement introduces much more versions which may (or may not) cause use of
+them.
+A possible solution here is implementing the cache as a LRU-like cache, making
+sure we evict the old entries when we reach a certain capacity.
+
+We should make sure we don't consume too much memory when fetching releases.
+It's especially critical when the user queries for the different versions, but
+might happen in other cases too.
+
+With the suggested solution we don't have a way to retain versions that were
+removed from the upgrade-channels API. While it's not probable we should keep
+that in mind.
+
+## Design Details
+
+### Open Questions
+
+* When implementing the binaries LRU cache, should we prioritize the latest
+  versions of each y-release? We expect those to be the most common ones, but
+  maybe it's better to let the LRU mechanism determine what are the common
+  versions by itself?
+* How should we limit the amount of versions handled? In case of an enormous
+  amount of version entries returned from the API, can we make sure to mitigate
+  the issue on our side?
```

Comment:
Not sure I understand this. Limit in what way? Isn't the idea behind this proposal that we don't limit what is available? What actual problem is this referencing?

Reply:
limit the amount of versions, for example by taking only the latest X versions in a stream
everything that has the potential to grow too much might cause issues in some ways, but just as an example it might consume too much memory, cause performance issues, etc

We need to see how it goes in the future, some things are hard to foresee

Reply:
Sure, but open questions are meant to be answered :sweat_smile: 

How many versions would we end up with if we adopted this approach today? Maybe try to configure all of them manually with the process we currently have and see if we run into any issues?

Reply:
Yeah, I did this calculation. See here: https://github.com/openshift/assisted-service/blob/4b4c27b5b99a3f1639f37cb81fb9ccd92a112f5f/docs/enhancements/dynamic-fetching-of-openshift-versions.md#L61

I can give the manual test a go, but my original concern is not about the *current* amount of images returned from the API but rather a potentially large response we're getting from it *in the future*. Do you understand? :)

Reply:
Sure, but is this something you want to address? Otherwise I think it should probably go into the risks/mitigations section.

Reply:
No, I think it's a minor one
I'll add it to the risks/mitigations section